tokenserver/assignment/sqlnode/sql.py

```diff
@@ -65,14 +65,18 @@
 """)
 
 
-_UPDATE_GENERATION_NUMBER = sqltext("""\
+_UPDATE_USER_RECORD_IN_PLACE = sqltext("""\
 update
     users
 set
-    generation = :generation
+    generation = COALESCE(:generation, generation),
+    keys_changed_at = COALESCE(:keys_changed_at, keys_changed_at)
 where
     service = :service and email = :email and
-    generation < :generation and replaced_at is null
+    generation <= COALESCE(:generation, generation) and
```

Comment:
I had to stare at these bits for a while. Keeping this logic in Python (probably around an actual sqlalchemy update expression) would help readability, but I'm fine committing it as is for now.

Reply:
This is mostly belt-and-braces stuff to ensure that we don't move any timestamps backwards, I'll add more of a comment in the statement in the meantime.